### PR TITLE
Welcome screen

### LINKED ISSUE #28: Add a welcome page to site
@nogainbar suggests that we could add a welcome of some sort. We are considering showing some simple text explaining what the App is, and only showing the buttons when the initial message is dismissed.

### PR DESCRIPTION
**Pull request by stevehopkinson**
Create a welcome screen that prompts the user to enter their name, then address the user by their name in subsequent conversations.

Fixes #28

### CODE REVIEW

File: src/components/WelcomeScreen.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+class WelcomeScreen extends React.Component {
+    constructor(props) {
+      super(props);
+      this.state = {value: ''};
+      this.handleSubmit = this.handleSubmit.bind(this);
+    }
+
+    handleSubmit (e) {
+      e.preventDefault();
+      console.log(this.input.value);
```
Comment (des-des) [inline src/components/WelcomeScreen.js]: :fire:

File: src/components/WelcomeScreen.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      this.props.setUsername(this.input.value);
```
Comment (des-des) [inline src/components/WelcomeScreen.js]: can we find this.input.value on the event obj? I do not think you need a ref here!

File: src/components/WelcomeScreen.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      <div className="welcomeScreen">
+        <img alt="Welcome cat" src="./cat.png" className="welcomeImage"/>
+        <p className="welcomeText">{"Hi! Great to meet you. I'm Sammy. What's your name?"}</p>
+        <form className="welcomeForm" onSubmit={this.handleSubmit}>
+          <input
+            type="text"
+            placeholder="Enter your name"
+            name="name"
+            className="welcomeInput"
+            ref={(input) => this.input = input} />
```
Comment (des-des) [inline src/components/WelcomeScreen.js:26]: see above

**Review by des-des — CHANGES_REQUESTED** [review]

File: src/components/WelcomeScreen.js (hunk 1 above)
Comment (stevehopkinson) [reply src/components/WelcomeScreen.js]: Is this a 'you're on fire!' fire, or a 'nuke it from orbit' fire?

File: src/components/WelcomeScreen.js (hunk 2 above)
Comment (stevehopkinson) [reply src/components/WelcomeScreen.js]: Nope. If you console.log the event object, it's a horrible React proxy object. You have to work horrible React witchcraft to make it work.

File: src/components/WelcomeScreen.js (hunk 3 above)
Comment (stevehopkinson) [reply src/components/WelcomeScreen.js:26]: Pull the branch down and try it without the ref - it unleashes unspeakable horrors.

**Comment by stevehopkinson** [comment]
This is all in keeping with Facebook's own guides for accessing user input through uncontrolled components: https://facebook.github.io/react/docs/uncontrolled-components.html

Apart from the console.log, obvs.

File: src/components/WelcomeScreen.js (hunk 1 above)
Comment (des-des) [reply src/components/WelcomeScreen.js]: delete it.. Sorry, prob not very clear

File: src/components/WelcomeScreen.js (hunk 2 above)
Comment (des-des) [reply src/components/WelcomeScreen.js]: ah, ok I am being thick, this value should actually be in the props for your component, ie the value of this input should live inside redux state.. Its a massive pain / problem with react / redux so we can leave it for now.

**Comment by stevehopkinson** [comment]
Console log deleted - should be good to go.

**Comment by des-des** [comment]
:+1:

**Review by des-des — APPROVED** [review]

**Comment by des-des** [comment]
As far as I am concerned forms + react + redux is a massive pain.

image

The idea with redux is that you do not keep the state of the input box in the dom, you keep it in the redux state and bind the onChange event to an action to keep it synced up.

The halfway approach is to keep the inputs value in the state of the component, again using onChange event to update the state of the component, and setting the value of the input explicitly from the components state

**Comment by des-des** [comment]
@stevehopkinson Happy to merge, re-assign when you have read my comment (just so I know you know)

**Comment by des-des** [comment]
Conflicts!

**Comment by stevehopkinson** [comment]
@des-des Merge conflicts fixed - should be ready to go.

**Pull request merged**